src/jrd/cch.cpp

```diff
@@ -3539,164 +3571,194 @@ static bool expand_buffers(thread_db* tdbb, ULONG number)
-		{
-			BufferDesc* bdb = BLOCK(que_inst, BufferDesc, bdb_que);
-			if (bdb->bdb_page == page || bdb->bdb_pending_page == page)
-				return bdb;
-		}
+		lruSync.unlock();
+		lruSync.lock(SYNC_EXCLUSIVE);
+		requeueRecentlyUsed(bcb);
 	}
+	else
+		bcb->bcb_flags &= ~BCB_free_pending;
 
 	return NULL;
 }
 
 
-static LatchState latch_buffer(thread_db* tdbb, Sync &bcbSync, BufferDesc *bdb,
-							   const PageNumber page, SyncType syncType, int wait)
+static BufferDesc* get_oldest_buffer(thread_db* tdbb, BufferControl* bcb)
 {
-	//++bdb->bdb_use_count;
+/**************************************
+ * Function description:
+ *       Get candidate for preemption
+ *       Found page buffer must have SYNC_EXCLUSIVE lock.
+ **************************************/
 
-	if (!(bdb->bdb_flags & BDB_free_pending)
-#ifdef SUPERSERVER_V2
-		&& (page != HEADER_PAGE_NUMBER)
-#endif
-		)
+	int walk = bcb->bcb_free_minimum;
+	BufferDesc* bdb = nullptr;
+
+	Sync lruSync(&bcb->bcb_syncLRU, FB_FUNCTION);
+	if (bcb->bcb_lru_chain.load() != NULL)
 	{
-		recentlyUsed(bdb);
+		lruSync.lock(SYNC_EXCLUSIVE);
+		requeueRecentlyUsed(bcb);
+		lruSync.downgrade(SYNC_SHARED);
 	}
+	else
+		lruSync.lock(SYNC_SHARED);
 
-	// If buffer is currently replacing by another page but still writting
-	// to disk we should wait until this write finished, else we could
-	// allocate another buffer and read old page image (or even zero's)
-	// from disk into new buffer
+	for (QUE que_inst = bcb->bcb_in_use.que_backward;
+		 que_inst != &bcb->bcb_in_use;
+		 que_inst = que_inst->que_backward)
+	{
+		bdb = nullptr;
 
-	const bool waitPending = ((bdb->bdb_flags & BDB_free_pending) && bdb->bdb_page == page);
+		// get the oldest buffer as the least recently used -- note
+		// that since there are no empty buffers this queue cannot be empty
 
-	bcbSync.unlock();
+		if (bcb->bcb_in_use.que_forward == &bcb->bcb_in_use)
+			BUGCHECK(213);	// msg 213 insufficient cache size
 
-	if (waitPending)
-	{
-		//--bdb->bdb_use_count;
-		if (wait == 0)
-			return lsTimeout;  // go out
+		BufferDesc* oldest = BLOCK(que_inst, BufferDesc, bdb_in_use);
 
-		Thread::yield();
-	}
-	else
-	{
-		const bool latchOk = bdb->addRef(tdbb, syncType, wait);
+		if (oldest->bdb_flags & BDB_lru_chained)
+			continue;
 
-		//--bdb->bdb_use_count;
+		if (oldest->bdb_use_count || !oldest->addRefConditional(tdbb, SYNC_EXCLUSIVE))
+			continue;
 
-		if (!latchOk)
-			return lsTimeout; // go out
+		/*if (!writeable(oldest))
+		{
+			oldest->release(tdbb, true);
+			continue;
+		}*/
 
-		if (bdb->bdb_page == page)
+		bdb = oldest;
+		if (!(bdb->bdb_flags & (BDB_dirty | BDB_db_dirty)) || !walk)
+			break;
+
+		if (!(bcb->bcb_flags & BCB_cache_writer))
+			break;
+
+		bcb->bcb_flags |= BCB_free_pending;
+		if (!(bcb->bcb_flags & BCB_writer_active))
+			bcb->bcb_writer_sem.release();
+
+		if (walk)
 		{
-			//bdb->bdb_flags &= ~(BDB_faked | BDB_prefetch);
-			return lsOk;
+			bdb->release(tdbb, true);
+//			if (!--walk)
```

Comment:
Why maintain commented code (and using wrong indentation)?

Reply:
Fixed